psycopg/psycopg/connection_async.py

```diff
@@ -319,34 +319,36 @@ async def notifies(
 
         nreceived = 0
 
-        while True:
-            # Collect notifications. Also get the connection encoding if any
-            # notification is received to makes sure that they are consistent.
-            try:
-                async with self.lock:
-                    ns = await self.wait(notifies(self.pgconn), timeout=timeout)
-                    if ns:
-                        enc = pgconn_encoding(self.pgconn)
-            except e._NO_TRACEBACK as ex:
-                raise ex.with_traceback(None)
-
-            # Emit the notifications received.
-            for pgn in ns:
-                n = Notify(pgn.relname.decode(enc), pgn.extra.decode(enc), pgn.be_pid)
-                yield n
-                nreceived += 1
-
-            # Stop if we have received enough notifications.
-            if stop_after is not None and nreceived >= stop_after:
-                break
-
-            # Check the deadline after the loop to ensure that timeout=0
-            # polls at least once.
-            if deadline:
-                timeout = min(_WAIT_INTERVAL, deadline - monotonic())
-                if timeout < 0.0:
+        ns: deque[Notify] = deque()
+        self.add_notify_handler(ns.append)
+        try:
+            while True:
+                # Collect notifications. Also get the connection encoding if any
+                # notification is received to makes sure that they are consistent.
```

Comment:
This comment is now stale: we don't get the encoding anymore here.

It would be also worth mentioning the reason why we want to generate notification in this slightly roundabout way, receiving them from handler rather than from as return value.

```suggestion
                # Collect notifications. Also get the connection encoding if any
                # notification is received to makes sure that they are consistent.
                # Collect notifications. Note that we don't use the return
                # value of the generator but we rely on the handler in order
                # to collect notifications that might have been generated and
                # consumed by another concurrent thread.
```

Reply:
Resolving this comment because it's been supplanted by more recent changes.